File: RB101_RB109_small_problems/medium_2/unlucky_days.rb
=begin

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: integer representing a year
-  Output: integer representing the number of Friday the 13ths in the given year

---

**Problem Domain:**

A new Date object can be created with:
Date.new(yyyy, mm, dd)

A Date can be checked if it's a friday with Date#friday?

---

**Explicit Requirements:**
- Assume the year is greater than 1752 (when the UK adopted the modern Gregorian Calendar) and the modern Gregorian Calendar will remain in use for the foreseeable future.
---


**Implicit Requirements:**
- Method should be named #friday_13th
---

**Clarifying Questions:**

1. None yet.

---

**Mental Model:**
Keep a counter of how many Friday the 13ths exist in the year.
Iterate from 1 to 12 (number of months in the year). In each iteration, create a Date class with the year argument, the month iteration, and the 13th day of the month. Check if the date is a Friday. If so, increment the counter by 1.
Return the counter.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

friday_13th(2015) == 3
friday_13th(1986) == 1
friday_13th(2019) == 2

---

Data Structure
--------------

---

Algorithm
---------

Code
----

=end

require 'date'

def friday_13th(year)
  count = 0
  1.upto(12) do |month|
    thirteenth = Date.new(year, month, 13)
    count += 1 if thirteenth.friday?
  end
  count
end

p friday_13th(2015) == 3
p friday_13th(1986) == 1
p friday_13th(2019) == 2

#Further Exploration
#Count the number of months that have five Fridays

def months_with_five_fridays(year)
  month_count = 0
  current_month = 1
  loop do
    date = Date.new(year, current_month, 1)
    friday_count = 0
    loop do
      friday_count += 1 if date.friday?
      date = date.next_day
      break if current_month != date.mon
    end
    month_count += 1 if friday_count >= 5
    current_month += 1
    break if current_month > 12
  end
  month_count
end

p months_with_five_fridays(2027)

File: JS210_small_problems/medium_2/unlucky_days.js
"use strict";
/*

Write a function that takes a year as an argument and returns the number of 'Friday the 13ths' in that year. 
You may assume that the year is greater than 1752 (when the modern Gregorian Calendar was adopted by the United Kingdom). 
You may also assume that the same calendar will remain in use for the foreseeable future.

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: an integer representing a year
-  Output: return an integer representing how many 'Friday the 13ths' occur in that year.

---

**Problem Domain:**

Nothing specific.

---

**Explicit Requirements:**
- Assume that the input year is greater than 1752
- Assume the same calendar will remain in use for the forseeable future.
---


**Implicit Requirements:**
- The function should be named fridayThe13ths
- The function should return an integer
---

**Clarifying Questions:**

1. Will there always be a single argument provided to the function?
  yes
2. Will the argument always be an integer of the Number type? (i.e. no BigInt types)
  Yes
3. Can we use built in Date methods to solve the problem?
  Yes
4. Will the input year always have only 4 digits?
  Yes.

---

**Mental Model:**

- Declare and initialize a count variable to 0. This will keep track of how many Friday the 13ths occur in the input year.
- Iterate over the months from 1 - 12.
  - Check if the 13th of the month is a Friday. If so, increment the count variable.
- Return the count variable's value.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

fridayThe13ths(1986);      // 1
fridayThe13ths(2015);      // 3
fridayThe13ths(2017);      // 2
fridayThe13ths(2023);      // 2

---

Data Structure
--------------

Use a string to "build" the date, like so:
"YYYY-MM-13T00:00:00"

Convert the string above to a Date object, then we can check if that Date is a Friday.

---

Algorithm
---------

- Declare and initialize a count variable to 0. This will keep track of how many Friday the 13ths occur in the input year.
- Iterate over the months from 1 - 12.
  - Check if the 13th of the month is a Friday. If so, increment the count variable.
- Return the count variable's value.

Code
----

*/

function fridayThe13ths(year) {
  const JANUARY = 0;
  const DECEMBER = 11;
  const FRIDAY = 5;
  
  let count = 0;

  for (let month = JANUARY; month < DECEMBER; month += 1) {
    let date = new Date(year, month, 13);
    if (date.getDay() === FRIDAY) count += 1;
  }

  return count;
}

fridayThe13ths(1986);      // 1
fridayThe13ths(2015);      // 3
fridayThe13ths(2017);      // 2
fridayThe13ths(2023);      // 2